File: src/notifier.py
"""
Notification Dispatcher + Channel Adapters (Telegram, Discord)
"""
import logging
from abc import ABC, abstractmethod

import requests

from src.config import Config
from src.models import Alert

logger = logging.getLogger(__name__)


class ChannelAdapter(ABC):
    """Interface chung cho tất cả kênh thông báo"""

    @abstractmethod
    def send(self, message: str) -> bool:
        """Gửi tin nhắn. Return True nếu thành công."""
        pass

    @abstractmethod
    def name(self) -> str:
        pass


class TelegramAdapter(ChannelAdapter):
    """Gửi tin nhắn qua Telegram Bot API"""

    API_URL = "https://api.telegram.org/bot{token}/sendMessage"

    def __init__(self, bot_token: str, chat_id: str):
        self.bot_token = bot_token
        self.chat_id = chat_id

    def name(self) -> str:
        return "Telegram"

    def send(self, message: str) -> bool:
        if not self.bot_token or not self.chat_id:
            logger.warning("Telegram: bot_token or chat_id not configured, skipping")
            return False

        url = self.API_URL.format(token=self.bot_token)
        payload = {
            "chat_id": self.chat_id,
            "text": message,
            "parse_mode": "HTML",
        }

        try:
            resp = requests.post(url, json=payload, timeout=10)
            if resp.status_code == 200:
                logger.info("Telegram: message sent successfully")
                return True
            else:
                logger.error(f"Telegram error {resp.status_code}: {resp.text}")
                return False
        except requests.RequestException as e:
            logger.error(f"Telegram request failed: {e}")
            return False


class DiscordAdapter(ChannelAdapter):
    """Gửi tin nhắn qua Discord Webhook"""

    def __init__(self, webhook_url: str):
        self.webhook_url = webhook_url

    def name(self) -> str:
        return "Discord"

    def send(self, message: str) -> bool:
        if not self.webhook_url:
            logger.warning("Discord: webhook_url not configured, skipping")
            return False

        payload = {"content": message}

        try:
            resp = requests.post(self.webhook_url, json=payload, timeout=10)
            if resp.status_code in (200, 204):
                logger.info("Discord: message sent successfully")
                return True
            else:
                logger.error(f"Discord error {resp.status_code}: {resp.text}")
                return False
        except requests.RequestException as e:
            logger.error(f"Discord request failed: {e}")
            return False


class NotificationDispatcher:
    """Điều phối gửi message đến các kênh đã cấu hình"""

    def __init__(self, config: Config):
        self.config = config
        self._adapters: list[ChannelAdapter] = []
        self._init_adapters()

    def _init_adapters(self):
        """Khởi tạo adapters dựa trên config"""
        if "telegram" in self.config.channels:
            adapter = TelegramAdapter(
                bot_token=self.config.telegram_bot_token,
                chat_id=self.config.telegram_chat_id,
            )
            self._adapters.append(adapter)
            logger.info("Telegram adapter initialized")

        if "discord" in self.config.channels:
            adapter = DiscordAdapter(
                webhook_url=self.config.discord_webhook_url,
            )
            self._adapters.append(adapter)
            logger.info("Discord adapter initialized")

    def send_alert(self, alert: Alert):
        """Gửi alert đến tất cả kênh đã cấu hình"""
        message = alert.format_message()
        logger.info(f"Sending alert to {len(self._adapters)} channel(s)...")

        for adapter in self._adapters:
            try:
                success = adapter.send(message)
                if not success:
                    # Retry 1 lần
                    logger.info(f"Retrying {adapter.name()}...")
                    adapter.send(message)
            except Exception as e:
                logger.error(f"Failed to send via {adapter.name()}: {e}")

    def send_message(self, message: str):
        """Gửi tin nhắn tùy ý (dùng cho startup, error alerts...)"""
        for adapter in self._adapters:
            try:
                adapter.send(message)
            except Exception as e:
                logger.error(f"Failed to send via {adapter.name()}: {e}")

File: notifier/notifier.go
package notifier

import (
	"bytes"
	"encoding/json"
	"fmt"
	"log"
	"net/http"
	"strings"
	"time"

	"github.com/user/notibot/config"
	"github.com/user/notibot/models"
)

// ChannelAdapter interface for Telegram and Discord
type ChannelAdapter interface {
	Send(message string) bool
	Name() string
}

// TelegramAdapter implementation
type TelegramAdapter struct {
	botToken string
	chatID   string
}

func (t *TelegramAdapter) Name() string { return "Telegram" }
func (t *TelegramAdapter) Send(message string) bool {
	if t.botToken == "" || t.chatID == "" {
		log.Println("Telegram: bot_token or chat_id not configured, skipping")
		return false
	}

	url := fmt.Sprintf("https://api.telegram.org/bot%s/sendMessage", t.botToken)
	payload := map[string]string{
		"chat_id":    t.chatID,
		"text":       message,
		// No parse_mode by default so we don't need to escape < > &
		"parse_mode": "HTML",
	}

	for key, val := range payload {
		if key == "text" {
			// Minimal escaping to prevent telegram api from rejecting the message
			val = strings.ReplaceAll(val, "&", "&amp;")
			val = strings.ReplaceAll(val, "<", "&lt;")
			val = strings.ReplaceAll(val, ">", "&gt;")
			payload[key] = val
		}
	}


	jsonBody, _ := json.Marshal(payload)
	resp, err := http.Post(url, "application/json", bytes.NewBuffer(jsonBody))
	if err != nil {
		log.Printf("Telegram POST error: %v", err)
		return false
	}
	defer resp.Body.Close()

	if resp.StatusCode == 200 {
		return true
	} else if resp.StatusCode == 429 {
		// Rate limit
		var result map[string]interface{}
		json.NewDecoder(resp.Body).Decode(&result)
		retryAfter := float64(5) // default 5s
		if parameters, ok := result["parameters"].(map[string]interface{}); ok {
			if rValue, ok := parameters["retry_after"].(float64); ok {
				retryAfter = rValue
			}
		}
		log.Printf("Telegram rate limited, retrying after %.0fs", retryAfter)
		time.Sleep(time.Duration(retryAfter) * time.Second)
		// Retry once simply by making request again inline
		respRetry, errRetry := http.Post(url, "application/json", bytes.NewBuffer(jsonBody))
		if errRetry == nil && respRetry.StatusCode == 200 {
			respRetry.Body.Close()
			return true
		}
		if respRetry != nil {
			respRetry.Body.Close()
		}
		return false
	}

	var errorData map[string]interface{}
	json.NewDecoder(resp.Body).Decode(&errorData)
	log.Printf("Telegram error %d: %v", resp.StatusCode, errorData)
	return false
}

// DiscordAdapter implementation
type DiscordAdapter struct {
	webhookURL string
}

func (d *DiscordAdapter) Name() string { return "Discord" }
func (d *DiscordAdapter) Send(message string) bool {
	if d.webhookURL == "" {
		log.Println("Discord: webhook_url not configured, skipping")
		return false
	}

	payload := map[string]string{"content": message}
	jsonBody, _ := json.Marshal(payload)

	resp, err := http.Post(d.webhookURL, "application/json", bytes.NewBuffer(jsonBody))
	if err != nil {
		log.Printf("Discord POST error: %v", err)
		return false
	}
	defer resp.Body.Close()

	if resp.StatusCode == 200 || resp.StatusCode == 204 {
		return true
	}
	log.Printf("Discord error %d", resp.StatusCode)
	return false
}

// Stats tracking sent/failed notifications
type Stats struct {
	Sent   int
	Failed int
	Queued int
}

// Dispatcher coordinates sending messages through configured channels in background
type Dispatcher struct {
	cfg      *config.Config
	adapters []ChannelAdapter
	queue    chan string
	Stats    Stats
}

// NewDispatcher creates the notification dispatcher
func NewDispatcher(cfg *config.Config) *Dispatcher {
	d := &Dispatcher{
		cfg:   cfg,
		queue: make(chan string, 1000), // buffered channel for messages
	}

	for _, ch := range cfg.Channels {
		if ch == "telegram" {
			d.adapters = append(d.adapters, &TelegramAdapter{botToken: cfg.TelegramBotToken, chatID: cfg.TelegramChatID})
			log.Println("Telegram adapter initialized")
		} else if ch == "discord" {
			d.adapters = append(d.adapters, &DiscordAdapter{webhookURL: cfg.DiscordWebhookURL})
			log.Println("Discord adapter initialized")
		}
	}

	return d
}

// StartWorker initiates the background goroutine to process the queue
func (d *Dispatcher) StartWorker() {
	go func() {
		log.Println("Notification worker started (background goroutine)")
		for msg := range d.queue {
			d.Stats.Queued--
			for _, adapter := range d.adapters {
				success := adapter.Send(msg)
				if success {
					d.Stats.Sent++
				} else {
					log.Printf("Failed to send via %s. Retrying...", adapter.Name())
					// Retry 1x inline
					if adapter.Send(msg) {
						d.Stats.Sent++
					} else {
						d.Stats.Failed++
						log.Printf("Failed to send via %s after retry.", adapter.Name())
					}
				}
			}
		}
	}()
}

func (d *Dispatcher) SendAlert(alert models.Alert) {
	d.SendMessage(alert.FormatMessage())
}

// SendMessage adds a message to the channel
func (d *Dispatcher) SendMessage(message string) {
	d.Stats.Queued++
	select {
	case d.queue <- message:
		// added successfully
	default:
		log.Println("Warning: Notification queue is full, dropping message.")
		d.Stats.Queued--
		d.Stats.Failed++
	}
}

// SendMessageSync sends a message directly, blocking execution (mostly used on startup/shutdown)
func (d *Dispatcher) SendMessageSync(message string) {
	for _, adapter := range d.adapters {
		adapter.Send(message)
	}
}
